lib/index.js

```diff
@@ -144,6 +141,9 @@ export function configureUnifiedLanguageServer(
     rcName
   }
 ) {
+  /** @type {undefined|Array<string>} */
+  let workspaces
```

Comment:
Since these workspaces are unique, I think it’s better to use a set.

```js
// Somewhere on top

/**
 * Convert a file URL to a path and trim a trailing slash.
 *
 * @param {string} fileUrl
 * @returns {string}
 */
function trimFileUrl(fileUrl) {
  return fileURLToPath(d.replace(/\/$/, ''))
}
```

Then here:

```js
/** @type Set<string> */
const workspaces = new Set()

/**
 * @param {string[]} fileUrls
 */
function addWorkspaces(fileUrls) {
  for (const fileUrl of fileUrls) {
    workspaces.add(trimFileUrl(fileUrl))
  }
}

/**
 * @param {string[]} fileUrls
 */
function removeWorkspaces(fileUrls) {
  for (const fileUrl of fileUrls) {
    workspaces.remove(trimFileUrl(fileUrl))
  }
}
```

Not using suggestions because multiple places are affected.

Reply:
3 places use file URLs, 1 place uses paths; we now use files. It becomes verbose to switch to using paths and having to use `trimFileUrl` in 3 places?

Reply:
I mainly thought it would be nice to handle trimming the trailing slash and deduplication when adding / removing the workspace. Although maybe that should actually explicitly not happen? I’m not sure anymore.

```js
// LSP uses `file:` URLs (hrefs), `unified-engine` expects a paths.
// `process.cwd()` does not add a final slash, but `file:` URLs often do.
```

This comment states `file:` URLs **often** contain a final slash. This implies they might not contain them sometimes. Is that so? That’s a bit confusing to me.

Reply:
Yes, trailing slashes are “optional”.

To explain: things are different when a) resolving with a trailing `/`, and b) resolving without one. 
Users could pass both. `new URL('.', import.meta.url)`, the pattern used in the tests to get the “dirname” of a URL to a file, results in a trailing slash (and it isn’t even using `./`, which more logically would add a slash).
But passing a trailing `/` as a CWD to the engine has different (broken) results compared to no `/`.

That’s why I remove it, optionally, when it exists.

Reply:
But where to do it: most of the code deals with LSP. Many functions deal with URLs.
Only one function deals with the engine, and with paths.

We now keep URLs in memory, and convert them in that single function that needs paths.
That’s simpler in code/in my head that converting them in 3 functions?

Reply:
At first I also thought it was stylistic, but this actually changes behaviour.

Let’s say the client sends the following requests:

1. add workspace `/myWorkspace/`
2. add workspace `/myWorkspace`
3. remove workspace `/myWorkspace`

In the current situation this would result in the workspaces set containing the single item `/myWorkspace/`.

If the trailing slash is removed when adding / removing them as I proposed, the workspaces set would be empty.

I don’t know which is better, but I do think this should be a conscious choice.

Reply:
> In the current situation this would result in the workspaces set containing the single item /myWorkspace/.

This sounds much much much better to me. Add 2 different workspaces, remove 1 — makes sense to me that the other is left?

Reply:
If you think that’s better, that’s ok with me. I just really don’t know. It’s an edge case  I think isn’t likely to occur anyway. :shrug: 

Approved!